packages/account/src/BiconomySmartAccount.ts

```diff
@@ -74,38 +75,89 @@ export class BiconomySmartAccount extends SmartAccount implements IBiconomySmart
+      this.owner = await this.signer.getAddress()
       this.chainId = await this.provider.getNetwork().then((net) => net.chainId)
+      await this.setAccountIndex(accountIndex)
+      this.isInited = true
     } catch (error) {
-      console.error(`Failed to call init: ${error}`);
+      Logger.error(`Failed to call init: ${error}`);
       throw error
     }
 
     return this
   }
 
+  private isInitialized(): boolean{
+    if (!this.isInited)
+    throw new Error('BiconomySmartAccount is not initialized. Please call BiconomySmartAccount.init() before interacting with any other function')
+    return true
+  }
+
+  private setProxyContractState(){
+    const proxyInstanceDto = {
+      smartAccountType: SmartAccountType.BICONOMY,
+      version: BICONOMY_IMPLEMENTATION_ADDRESSES[this.address],
+      contractAddress: this.address,
+      provider: this.provider
+    }
+    this.proxy = getProxyContractInstance(proxyInstanceDto)
+  }
+
+  private setEntryPointContractState(){
+    const _entryPointAddress = this.smartAccountInfo.entryPointAddress
+    this.setEntrypointAddress(_entryPointAddress)
+    const entryPointInstanceDto = {
+      smartAccountType: SmartAccountType.BICONOMY,
+      version: ENTRYPOINT_ADDRESSES[_entryPointAddress],
+      contractAddress: _entryPointAddress,
+      provider: this.provider
+    }
+    this.entryPoint = getEntryPointContractInstance(entryPointInstanceDto)
+  }
+
+  private setFactoryContractState(){
+    const _factoryAddress = this.smartAccountInfo.factoryAddress
+    const factoryInstanceDto = {
+      smartAccountType: SmartAccountType.BICONOMY,
+      version: BICONOMY_FACTORY_ADDRESSES[_factoryAddress],
+      contractAddress: _factoryAddress,
+      provider: this.provider
+    }
+    this.factory = getFactoryContractInstance(factoryInstanceDto)
+  }
+
+  private async setContractsState() {
+    this.setProxyContractState()
+    this.setEntryPointContractState()
+    this.setFactoryContractState()
+  }
+
   async setAccountIndex(accountIndex: number): Promise<void> {
     this.accountIndex = accountIndex
     this.address = await this.getSmartAccountAddress(accountIndex)
-    this.proxy = await SmartAccount_v100__factory.connect(this.address, this.provider)
+    await this.setContractsState()
+    await this.setInitCode(this.accountIndex)
   }
 
   async getSmartAccountAddress(accountIndex: number = 0): Promise<string> {
     try {
       this.isSignerDefined()
-      await this.getInitCode(accountIndex)
-      const address = await this.factory.getAddressForCounterFactualAccount(
-        await this.signer.getAddress(),
-        ethers.BigNumber.from(accountIndex)
-      )
-      return address
+      let smartAccountsList: ISmartAccount[] = (await this.getSmartAccountsByOwner({
+        chainId: this.chainId,
+        owner: this.owner
+      })).data
+      smartAccountsList = smartAccountsList.filter((smartAccount: ISmartAccount) => { return accountIndex === smartAccount.index })
```

Comment:
Do proper error handling here. Handle the case if smartAccountList is undefined.

Reply:
there is already a check in place. filter function returns array so undefined case don't need to be managed
if (smartAccountsList.length === 0)
        throw new Error('Failed to get smart account address')

Reply:
I'm saying what if smartAccountsList is null

Reply:
That is not going to happen, as the backend API ensures the return of the account address. just in case added this
`
 if ( !smartAccountsList )
      throw new Error('Failed to get smart account address. Please raise an issue on https://github.com/bcnmy/biconomy-client-sdk for further investigation.')
`